drivers/nrf_wifi/bus_if/bal/inc/bal_ops.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/**
+ * @brief Header containing the OPs declarations for the Bus Abstraction Layer
+ * (BAL) of the Wi-Fi driver.
+ */
+
+#ifndef __BAL_OPS_H__
+#define __BAL_OPS_H__
+
+/**
+ * struct nrf_wifi_bal_ops - Ops to be provided by a particular bus
+ *                           implementation.
+ * @init:
+ * @deinit:
+ * @dev_init:
+ * @dev_deinit:
+ * @read_word:
+ * @write_word:
+ * @read_block:
+ * @write_block:
+ * @dma_map:
+ * @dma_unmap:
+ */
+struct nrf_wifi_bal_ops {
+	void * (*init)(struct nrf_wifi_osal_priv *opriv,
+		       void *cfg_params,
+		       enum nrf_wifi_status (*intr_callbk_fn)(void *hal_ctx));
+	void (*deinit)(void *bus_priv);
+	void * (*dev_add)(void *bus_priv,
+			  void *bal_dev_ctx);
+	void (*dev_rem)(void *bus_dev_ctx);
+
+	enum nrf_wifi_status (*dev_init)(void *bus_dev_ctx);
+	void (*dev_deinit)(void *bus_dev_ctx);
+	unsigned int (*read_word)(void *bus_dev_ctx,
+				  unsigned long addr_offset);
+	void (*write_word)(void *bus_dev_ctx,
+			   unsigned long addr_offset,
+			   unsigned int val);
+	void (*read_block)(void *bus_dev_ctx,
+			   void *dest_addr,
+			   unsigned long src_addr_offset,
+			   size_t len);
+	void (*write_block)(void *bus_dev_ctx,
+			    unsigned long dest_addr_offset,
+			    const void *src_addr,
+			    size_t len);
+	unsigned long (*dma_map)(void *bus_dev_ctx,
+				 unsigned long virt_addr,
+				 size_t len,
+				 enum nrf_wifi_osal_dma_dir dma_dir);
+	unsigned long (*dma_unmap)(void *bus_dev_ctx,
+				   unsigned long phy_addr,
+				   size_t len,
+				   enum nrf_wifi_osal_dma_dir dma_dir);
+#ifdef CONFIG_NRF_WIFI_LOW_POWER
+	void (*rpu_ps_sleep)(void *bus_dev_ctx);
+	void (*rpu_ps_wake)(void *bus_dev_ctx);
+	int (*rpu_ps_status)(void *bus_dev_ctx);
+#endif /* CONFIG_NRF_WIFI_LOW_POWER */
```

Comment:
isn't this tied to Kconfig, ie, Zephyr?

Reply:
Not really, e.g., in Linux driver these are passed as cflags and can be based on Linux configurations.

Reply:
HAL code should avoid using such pattern, CMake can inject necessary definitions based on Zephyr's Kconfig, see `modules/hal_nordic` in Zephyr's tree.

Reply:
> CMake can inject necessary definitions based on Zephyr's Kconfig

I didn't quite understand, even if Zephyr CMake passes the definitions, these would still need to used, no?
I looked at https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/hal_nordic/nrf_802154/CMakeLists.txt#L34, are you suggesting to drop `CONFIG_` prefix and then translate Zephyr to Agnostic defines? I can do that, but IMHO the prefix is pretty standard across OSes.

Reply:
Translated the Kconfig options.